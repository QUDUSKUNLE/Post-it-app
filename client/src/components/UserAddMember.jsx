import React from 'react';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import { Redirect } from 'react-router-dom';
import { addMember, getAllUsers } from '../actions/memberActions';
import MemberStore from '../stores/MemberStore';
import GroupStore from '../stores/GroupStore';
import { getUserGroups } from '../actions/GroupActions';

/**
 * @export
 * @description - renders AddMember Component
 * @class UserAddMember
 * @extends {React.Component}
 */
export default class UserAddMember extends React.Component {

  /**
   * Creates an instance of UserAddMember.
   * @constructor
   * @param {*} props -
   * @memberof UserAddMember
   */
  constructor(props) {
    super(props);
    this.state = {
      loggedIn: JSON.parse(localStorage.getItem('userIn')),
      userId: JSON.parse(localStorage.getItem('Id')),
      groups: [],
      registeredUsers: [],
      group: {},
      member: ''
    };
    /**
     * @description This binding is necessary to make `this` work
     * in the callback
     */
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.handleAddMemberToGroup = this.handleAddMemberToGroup.bind(this);
    this.userGroups = this.userGroups.bind(this);
  }

  /**
   * @memberof UserAddMember
   * @return {*} void
   */
  componentWillMount() {
    getUserGroups(this.state.userId);
    getAllUsers();
  }

  /**
   * @memberof UserAddMember
   * @return {*} void
   */
  componentDidMount() {
    GroupStore.on('GET_USER_GROUPS', this.userGroups);
    MemberStore.on('ALL_USERS', this.userGroups);
    MemberStore.on('ADD_MEMBER', this.handleAddMemberToGroup);
  }

  componentWillUnmount() {
    GroupStore.removeListener('GET_USER_GROUPS', this.userGroups);
    MemberStore.removeListener('ALL_USERS', this.userGroups);
    MemberStore.removeListener('ADD_MEMBER', this.handleAddMemberToGroup);
  }

  /**
   * @memberof UserAddMember
   * @return {*} void
   */
  userGroups() {
    const allUsers = MemberStore.registeredUsers();
    this.setState({
      groups: GroupStore.allGroups(),
      registeredUsers: allUsers.filter(user =>
        (Object.keys(user)[0] !== this.state.userId))
    });
  }
  /**
   * @description - onChange event
   * @param {event} event - event
   * @returns {null} null
   * @memberOf ChatBox
   */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  /**
	 * @description This handles addMember form submission
	 * @param {object} event .
	 * @returns {void} .
	 */
  onSubmit(event) {
    event.preventDefault();
    const groupDetails = this.state.group.split(',');
    const memberDetails = {
      groupId: groupDetails[0],
      group: groupDetails[1],
      memberId: this.state.member
    };
    addMember(memberDetails);
  }

  /**
	 * @description This handles addMember update from store
	 * @param {object} event .
	 * @returns {void} .
	 */
  handleAddMemberToGroup() {
    const addMemberResponse = MemberStore.addMember();
    toastr.success(addMemberResponse);
    this.props.history.push('/broadcastboard');
  }

  /**
   * @description - render method, React lifecycle method
   * @returns {Object} AddMember component
   * @AddMember
   */
  render() {
    if (!this.state.loggedIn) {
      return (
        <Redirect to="/signin" />
      );
    }

    return (
      <div className="container addmember">
        <div className="row">
          <div className="col-md-offset-3 col-md-6">
            <div className="row w3-card w3-white">
              <form className="addmemberform" onSubmit={this.onSubmit}>
                <div className="form-group">
                  <label htmlFor="groupname">Group Name</label>
                  <select name="group" onChange={this.onChange}
                    className="form-control">
                    <option value="">Select a group</option>
                    {(this.state.groups).map(group =>
                      <option key={Object.values(group)}
                        value={[Object.values(group), Object.keys(group)]}>
                        {Object.keys(group)}</option>
                    )}
                  </select>
                </div>
                <div className="form-group">
                  <label htmlFor="members">Member</label>
                  <select name="member" onChange={this.onChange}
                    className="form-control">
                    <option value="">add member to group</option>
                    {(this.state.registeredUsers).map(user =>
                      <option key={Object.keys(user)}
                        value={Object.keys(user)}>
                        {Object.values(user)}</option>
                    )}
                  </select>
                </div>
                <button type="submit" className="signinformbtn">Add Member
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
  );
  }
}

// props validation
UserAddMember.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  })
};
