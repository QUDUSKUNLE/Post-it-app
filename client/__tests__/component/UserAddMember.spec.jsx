import React from 'react';
import expect from 'expect';
import PropTypes from 'prop-types';
import { mount } from 'enzyme';
import sinon from 'sinon';
import UserAddMember from '../../src/components/UserAddMember';
import localStorageMock from '../../src/__mock__/localStorage';

window.localStorage = localStorageMock;
describe('<UserAddMember/>', () => {
  const props = {
    groups: [],
    registeredUsers: [],
    group: {},
    member: '',
    closeModal: jest.fn()
  };
  const wrapper = mount(<UserAddMember {...props} />,
    {
      childContextTypes: { router: PropTypes.object },
      context: {
        router:
        {
          history: {
            push: () => null,
            replace: () => null,
            createHref: () => null,
            path: '/member',
            component: '[function AddMember]',
            location: {
              pathname: '/member',
              search: '',
              hash: '',
              key: 'ukxugb'
            },
            computedMatch: {
              path: '/member',
              url: '/member',
              isExact: true,
              params: {}
            }
          }
        }
      }
    }
  );

  it('component is expected to be defined', () => {
    expect(UserAddMember).toBeDefined();
    expect(wrapper.state()).toBeDefined();
  });

  it('component methods expected to be defined', () => {
    expect(wrapper.nodes[0].handleAddMemberToGroup).toBeDefined();
  });

  it('should call componentDidMount component lifecycle', () => {
    const spy = sinon.spy(UserAddMember.prototype, 'componentDidMount');
    wrapper.instance().componentDidMount();
    expect(spy.calledOnce).toBeTruthy();
  });

  it('should call componentWillUnmount component lifecycle', () => {
    const spy = sinon.spy(UserAddMember.prototype, 'componentWillUnmount');
    wrapper.instance().componentWillUnmount();
    expect(spy.calledOnce).toBeTruthy();
  });

  it('component should call onChange method', () => {
    const event = { target: { name: 'name', value: 'value' } };
    wrapper.instance().onChange(event);
    expect(wrapper.state().name).toEqual('value');
  });
});
