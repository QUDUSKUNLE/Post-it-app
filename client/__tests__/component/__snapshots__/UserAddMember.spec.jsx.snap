// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserAddMember/> component should render correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="container addmember"
  >
    <div
      className="row"
    >
      <div
        className="col-md-offset-3 col-md-6"
      >
        <h5
          className="text-center"
        >
          <b>
            Add Member to group
          </b>
        </h5>
        <div
          className="row w3-card w3-white"
        >
          <form
            className="addmemberform"
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <label
                htmlFor="groupname"
              >
                Group Name
              </label>
              <select
                className="form-control"
                name="group"
                onChange={[Function]}
              >
                <option
                  value=""
                >
                  Select a group
                </option>
              </select>
            </div>
            <div
              className="form-group"
            >
              <label
                htmlFor="members"
              >
                Member
              </label>
              <select
                className="form-control"
                name="member"
                onChange={[Function]}
              >
                <option
                  value=""
                >
                  add member to group
                </option>
              </select>
            </div>
            <button
              className="signinformbtn"
              type="submit"
            >
              Add Member
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "nodes": Array [
    <div
      className="container addmember"
    >
      <div
        className="row"
      >
        <div
          className="col-md-offset-3 col-md-6"
        >
          <h5
            className="text-center"
          >
            <b>
              Add Member to group
            </b>
          </h5>
          <div
            className="row w3-card w3-white"
          >
            <form
              className="addmemberform"
              onSubmit={[Function]}
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="groupname"
                >
                  Group Name
                </label>
                <select
                  className="form-control"
                  name="group"
                  onChange={[Function]}
                >
                  <option
                    value=""
                  >
                    Select a group
                  </option>
                </select>
              </div>
              <div
                className="form-group"
              >
                <label
                  htmlFor="members"
                >
                  Member
                </label>
                <select
                  className="form-control"
                  name="member"
                  onChange={[Function]}
                >
                  <option
                    value=""
                  >
                    add member to group
                  </option>
                </select>
              </div>
              <button
                className="signinformbtn"
                type="submit"
              >
                Add Member
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <UserAddMember
        group={Object {}}
        groups={Array []}
        loggedIn={true}
        member=""
        registeredUsers={Array []}
        userId="AZCVGFRTUINSMUY15156"
      />,
      "_debugID": 41,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": UserAddMember {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleAddMemberToGroup": [Function],
        "onChange": [Function],
        "onSubmit": [Function],
        "props": Object {
          "group": Object {},
          "groups": Array [],
          "loggedIn": true,
          "member": "",
          "registeredUsers": Array [],
          "userId": "AZCVGFRTUINSMUY15156",
        },
        "refs": Object {},
        "state": Object {
          "group": "",
          "groups": Array [],
          "loggedIn": true,
          "member": "",
          "registeredUsers": Array [],
          "userId": "AZCVGFRTUINSMUY15156",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
        "userGroups": [Function],
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="container addmember"
        >
          <div
            className="row"
          >
            <div
              className="col-md-offset-3 col-md-6"
            >
              <h5
                className="text-center"
              >
                <b>
                  Add Member to group
                </b>
              </h5>
              <div
                className="row w3-card w3-white"
              >
                <form
                  className="addmemberform"
                  onSubmit={[Function]}
                >
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="groupname"
                    >
                      Group Name
                    </label>
                    <select
                      className="form-control"
                      name="group"
                      onChange={[Function]}
                    >
                      <option
                        value=""
                      >
                        Select a group
                      </option>
                    </select>
                  </div>
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="members"
                    >
                      Member
                    </label>
                    <select
                      className="form-control"
                      name="member"
                      onChange={[Function]}
                    >
                      <option
                        value=""
                      >
                        add member to group
                      </option>
                    </select>
                  </div>
                  <button
                    className="signinformbtn"
                    type="submit"
                  >
                    Add Member
                  </button>
                </form>
              </div>
            </div>
          </div>
        </div>,
        "_debugID": 42,
        "_renderedOutput": <div
          className="container addmember"
        >
          <div
            className="row"
          >
            <div
              className="col-md-offset-3 col-md-6"
            >
              <h5
                className="text-center"
              >
                <b>
                  Add Member to group
                </b>
              </h5>
              <div
                className="row w3-card w3-white"
              >
                <form
                  className="addmemberform"
                  onSubmit={[Function]}
                >
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="groupname"
                    >
                      Group Name
                    </label>
                    <select
                      className="form-control"
                      name="group"
                      onChange={[Function]}
                    >
                      <option
                        value=""
                      >
                        Select a group
                      </option>
                    </select>
                  </div>
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="members"
                    >
                      Member
                    </label>
                    <select
                      className="form-control"
                      name="member"
                      onChange={[Function]}
                    >
                      <option
                        value=""
                      >
                        add member to group
                      </option>
                    </select>
                  </div>
                  <button
                    className="signinformbtn"
                    type="submit"
                  >
                    Add Member
                  </button>
                </form>
              </div>
            </div>
          </div>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <UserAddMember
    group={Object {}}
    groups={Array []}
    loggedIn={true}
    member=""
    registeredUsers={Array []}
    userId="AZCVGFRTUINSMUY15156"
  />,
}
`;
